crates/shared/src/price_estimation/competition.rs

```diff
@@ -203,12 +197,37 @@ impl PriceEstimating for RacingCompetitionEstimator<Arc<dyn PriceEstimating>> {
             query.clone(),
             query.kind,
             |estimator, query| estimator.estimate(query),
-            move |a, b, context| {
-                if is_second_quote_result_preferred(query.as_ref(), a, b, context) {
-                    Ordering::Less
-                } else {
-                    Ordering::Greater
-                }
+            move |results, context| {
+                let gas_costs = results
+                    .iter()
+                    .filter_map(|(_, r)| Some(r.as_ref().ok()?.gas))
+                    .sorted()
+                    .collect_vec();
+                let median_gas = *gas_costs.get(gas_costs.len() / 2usize).unwrap_or(&0);
+                let gas_lower_bound = std::cmp::max(1, median_gas / 2);
+
+                results
+                    .iter()
+                    .map(|(_, result)| result)
+                    .enumerate()
+                    // Solvers could try to game the system by providing unreasonably low gas amounts.
+                    // That way they'll win the best quote and the protocol will quote a very low fee
+                    // to the user (which is basically a huge subsidy we'd have to pay).
+                    // To avoid that we discard all quotes with gas costs significantly below the
+                    // median.
+                    .filter(|(_, r)| match r {
+                        Ok(estimate) => {
+                            let keep = estimate.gas >= gas_lower_bound;
+                            if !keep {
+                                tracing::warn!(?estimate, "discarding outlier estimate");
+                            }
+                            keep
+                        }
+                        Err(_) => true,
+                    })
+                    .max_by(|a, b| compare_quote_result(&query, a.1, b.1, context))
+                    .map(|(index, _)| index)
+                    .expect("we never discard all the results so at least 1 should remain")
```

Comment:
Does it make any sense to add a test case for this scenario just to make sure the logic rejects all the unexpected results not giving a chance to any?

Reply:
What do you mean by "not giving a chance to any"?
I added a test case that verifies that low gas costs estimates get rejected. Are you suggesting to add a test case where discarding solutions might lead to a problem?
The only thing that comes to mind is if all solutions report exactly the same gas cost. In that case a line like:
```rust
let keep = estimate.gas > gas_lower_bound;
```
(instead of `>=` like we have right now) would discard all solutions and trigger a panic.
Is that what you had in mind?

Reply:
Actually after looking more at the code we could in fact trigger this panic if all solutions report exactly 0 gas. 🤔

Reply:
> Actually after looking more at the code we could in fact trigger this panic if all solutions report exactly 0 gas. 🤔
Yep, I am about this case